File: linkedlist/LL-cycle.cpp
// BRUTE force
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        unordered_map<ListNode*,int> visited;
        ListNode *temp = head;
        
        while(temp!=NULL){
            if(visited.find(temp)!=visited.end()){
                return true;
            }visited[temp]=1;
            temp=temp->next;
            
        }
        return false;
    }
};

// OPTIMAL :
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *fast = head;
        ListNode *slow = head;

        while(fast!=NULL && fast->next!=NULL){
            slow=slow->next;
            fast=fast->next->next;
            if(slow == fast){
                return true;
            }
            
        }
        return false;
    }
};

File: linkedlist/LL-cycle.java
// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; next = null; }
}

public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;           // Move slow pointer by 1 step
            fast = fast.next.next;      // Move fast pointer by 2 steps

            if (slow == fast) {         // If the two pointers meet, there is a cycle
                return true;
            }
        }
        
        return false;  // If no cycle is detected
    }
}